### PR TITLE
♻️ user services can no longer contact the dynamic sidecar

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

The dy-sidecar was accidentally expoed to the user services instead of only exposing the proxy.
It is no longer possible to contact the dy-sidecar from a user service (like from a jupyter-notebook).

## Related issue/s

- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3390

## Checklist

- [x] Unit tests for the changes exist

### CODE REVIEW

**Review by mrnicegyu11 — CHANGES_REQUESTED** [review]
Thanks a lot for this, the effort to fix this is MUCH appreciated from a devsecops point of view.

Would it be possible to add a test that runs a dynamic sidecar service, gets the IP (all relevant network interfaces IPs) of the dynamic-sidecar service and then assert that one cannot curl / send REST requests to the dynamic sidecar from the service's container using the various IPs of the dynamic sidecar?

Then we are safe for the future. If this is not feasible I guess the current changes to what is necessary to prevent the oSparc Simcore priviledge escalation and container escape.

In any case, I think in the mid-term it will be necessay that the dynamic sidecar does not just accept all REST API requests containing valid docker json and execute them, as this opens the door for further attacks.

One way to mitigate this would be (just from the top of my head) giving the dy-sidecars a nonce when starting, that only the director-v2 know, best only stored transiently in memory. Then only director-v2 can make the dy-sidecars execute docker commands. What are your thoughts?

**Review by sanderegg — APPROVED** [review]
looking good.
So now the user services are not in a common network with the dy-sidecar.
So now the only connection of the user services should be through the dy-proxy, then to simcore-traefik, then to ops-traefik and finally to the client user.
How is the dy-proxy network wise? only connected to traefik or are there more connections?
Also how does that work with the control loop experiment?

**Comment by GitHK** [comment]
> So now the user services are not in a common network with the dy-sidecar.

Correct.

> So now the only connection of the user services should be through the dy-proxy, then to simcore-traefik, then to ops-traefik and finally to the client user.

Yes that is my understanding 

> How is the dy-proxy network wise? only connected to traefik or are there more connections?

proxy only has the essential network to make it work `interactive_services_network` and the `dy_sidecar_...`. Looks OK.

> Also how does that work with the control loop experiment?

Yes uses a separate network

**Review by pcrespov — APPROVED** [review]

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py [hunk 1]
```diff
@@ -234,10 +233,7 @@ def get_dynamic_sidecar_spec(
             "service_image": dynamic_sidecar_settings.DYNAMIC_SIDECAR_IMAGE,
         },
         "name": scheduler_data.service_name,
-        "networks": [
-            {"Target": swarm_network_id},
-            {"Target": dynamic_sidecar_network_id},
-        ],
+        "networks": [{"Target": swarm_network_id}],
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py:236]: swarm_network_id is the id of the swarm_network, which is the network on which interactive services talk (should be something like: intearactive_services_network).
Why this name? I guess it's a historic thing.

**Review by mrnicegyu11 — APPROVED** [review]

**Comment by GitHK** [comment]
> Thanks a lot for this, the effort to fix this is MUCH appreciated from a devsecops point of view.
> 
> Would it be possible to add a test that runs a dynamic sidecar service, gets the IP (all relevant network interfaces IPs) of the dynamic-sidecar service and then assert that one cannot curl / send REST requests to the dynamic sidecar from the service's container using the various IPs of the dynamic sidecar?
> 
> Then we are safe for the future. If this is not feasible I guess the current changes to what is necessary to prevent the oSparc Simcore priviledge escalation and container escape.
> 
> In any case, I think in the mid-term it will be necessay that the dynamic sidecar does not just accept all REST API requests containing valid docker json and execute them, as this opens the door for further attacks.
> 
> One way to mitigate this would be (just from the top of my head) giving the dy-sidecars a nonce when starting, that only the director-v2 know, best only stored transiently in memory. Then only director-v2 can make the dy-sidecars execute docker commands. What are your thoughts?

As discussed the suggested test would be had to code maintain, run slow and get has the risk of becoming a false positive.

The idea of having a token to secure communication between director-v2 and dy-sidecar is a better option.

**Pull request merged**